Containerized Docker Application Lifecycle - 05.2 (#10619)
dotnet/docs: docs/standard/containerized-lifecycle-architecture/docker-devops-workflow/docker-application-outer-loop-devops-workflow.md

![Browser view of the Docker CI pipeline in Azure DevOps](./media/image3.png)

**Figure 5-3**. The Docker CI pipeline in Azure DevOps Services including Build & Release Templates and associated tasks.

You can use these templates and tasks to construct your CI/CD artifacts to Build / Test and Deploy in Azure Service Fabric, Azure Kubernetes Service, and similar offerings.

With these Visual Studio Team Services tasks, a build Linux-Docker Host/VM provisioned in Azure and your preferred Docker registry (Azure Container Registry, Docker Hub, private Docker DTR, or any other Docker registry) you can assemble your Docker CI pipeline in a very consistent way.

***Requirements:***

- Azure DevOps Services, or for on-premises installations, Team Foundation Server 2015 Update 3 or later.

- An Azure DevOps Services agent that has the Docker binaries.

  An easy way to create one of these agents is to use Docker to run a container based on the Azure DevOps Services agent Docker image.

> > [!INFORMATION]

Reviewer: INFORMATION is not one of the note types. Should this be a !NOTE instead?

https://github.com/dotnet/docs/blob/master/styleguide/template.md#alerts

Author: Yep, I just told @rpetrusha I'll include this as one more Regex change.

Following are several types of tests that you can run against Docker containers:

- Unit tests for Docker containers

- Testing groups of interrelated applications or microservices

- Test in production and "canary" releases

The important point is that when running integration and functional tests, you must run those tests from outside of the containers. Tests are not contained or run in the containers you are deploying, because the containers are based on static images that should be exactly like the ones you'll be deploying to production.

A practical option when testing more advanced scenarios, like including several clusters (test cluster, staging cluster, and production cluster) is to publish the images to a registry, so it can be tested in various clusters.

### Push the custom application Docker image into your global Docker Registry

> After the Docker images have been tested and validated, you'll want to tag and publish them to your Docker registry. The Docker registry is a critical piece in the Docker application life cycle because it is the central place where you store your custom test (aka "blessed images") to be deployed into QA and production environments.

Reviewer: aka is bad for loc 

```suggestion
After the Docker images have been tested and validated, you'll want to tag and publish them to your Docker registry. The Docker registry is a critical piece in the Docker application life cycle because it is the central place where you store your custom test (also known as "blessed images") to be deployed into QA and production environments.
```

Similar to how the application code stored in your SCC repository (Git, etc.) is your "source of truth," the Docker registry is your "source of truth" for your binary application or bits to be deployed to the QA or production environments.

> Typically, you might want to have your private repositories for your custom images either in a private repository in Azure Container Registry or in an on-premises registry like Docker Trusted Registry, or in a public-cloud registry with restricted access (like Docker Hub), although in this last case if your code is not open source, you must trust the vendor's security. Either way, the method you use is similar and is based on the `docker push` command, as depicted in Figure 5-4.

Reviewer: can we use less formal words?

```suggestion
Typically, you might want to have your private repositories for your custom images either in a private repository in Azure Container Registry or in an on-premises registry like Docker Trusted Registry, or in a public-cloud registry with restricted access (like Docker Hub), although in this last case if your code is not open source, you must trust the vendor's security. Either way, the method you use is similar and is based on the `docker push` command, as shown in Figure 5-4.
```

![In step 3, for building integration and testing (CI) you might publish the resulting docker images to a private or public registry.](./media/image4.png)

**Figure 5-4**. Publishing custom images to Docker Registry

There are multiple offerings of Docker registries from cloud vendors like Azure Container Registry, Amazon Web Services Container Registry, Google Container Registry, Quay Registry, and so on.

Using the Docker tasks, you can push a set of service images defined by a `docker-compose.yml` file, with multiple tags, to an authenticated Docker registry (like Azure Container Registry), as shown in Figure 5-5.

![Browser view of the step to publish images to a registry from Azure DevOps.](./media/image5.png)

**Figure 5-5**. Using Azure DevOps Services to publishing custom images to a Docker Registry

> > [!INFORMATION]

Reviewer: Same thing here

> > To learn more about Azure Container Registry, go to <https://aka.ms/azurecontainerregistry>.

Reviewer: Our boilerplate would be 

```suggestion
> For more information about Azure Container Registry, see <https://aka.ms/azurecontainerregistry>.
```

![Browser view of the task catalog in Azure DevOps, showing the Docker deploy task.](./media/image10.png)

**Figure 5-10**. Adding the Docker Deploy task to your Environment RM

Figure 5-11 demonstrates how you can edit the Docker Deploy task and specify the Target Type (Azure Container Service DC/OS, in this case), your Docker Compose File, and the Docker Registry connection (like Azure Container Registry or Docker Hub). This is the task that will retrieve your ready-to-use custom Docker images to be deployed as containers in the cluster.

![Browser view of Azure DevOps, deploy to orchestrator task definition.](./media/image11.png)

**Figure 5-11**. Docker Deploy task definition deploying to Azure Container Service DC/OS

> > [!INFORMATION]

Reviewer: same issue here and in the sentence about alert type and boilerplate

> To read more about the CD pipeline with Azure DevOps Services and Docker, visit <https://azure.microsoft.com/services/devops/pipelines>

## Step 5: Run and manage

Because running and managing applications at enterprise-production level is a major subject in and of itself, and due to the type of operations and people working at that level (IT operations) as well as the large scope of this area, we have devoted the entire next chapter to explaining it.

## Step 6: Monitor and diagnose

> This topic also is covered in the next chapter as part of the tasks that IT performs in production systems; however, is important to highlight that the insights obtained in this step must feed back to the development team so that the application is constantly improved. From that point of view, it is also part of DevOps, although the tasks and operations are commonly performed by IT.

Reviewer: Topic is legacy terminology 

```suggestion
This article also is covered in the next chapter as part of the tasks that IT performs in production systems; however, is important to highlight that the insights obtained in this step must feed back to the development team so that the application is constantly improved. From that point of view, it is also part of DevOps, although the tasks and operations are commonly performed by IT.
```

Author: @mairaw I'm not sure about the change from topic to article, because it refers to the "Monitor(ing) and diagnose(ing)" topic or subject, and that's what's covered in another chapter.

Reviewer 2: I agree; "topic" is correct in this context. I'll merge your PR now, @mvelosop.
